### DISCUSSION TITLE
Different requests to same client at same time using Flask Socketio

### DISCUSSION THREAD

**Discussion by athulvis** [miguelgrinberg/Flask-SocketIO#1931] [Q&A] [answered]
I am creating an offline live dashboard for updating data from client to server using Flask. I have multiple data to update and it works well if I send them in a single request. But I have two types of requests and one updates slower than another. Is there anyway to update the different requests simultaneously in the client?

what I tried: I created another function and run the code as below. But only first value is updating and second isn't.

what I need: Both test and test2 should be updated simultaneously with its own delay (test has no delay and test2 has 3 seconds delay).

**app.py:**

```from flask import Flask, render_template
from flask_socketio import SocketIO

app = Flask(__name__)
socketio = SocketIO(app, logger=True, engineio_logger=True)

@app.route('/')
def index():
    return render_template('index.html')

@socketio.on('check')
def gen(data):
    j=0
    while True:
        socketio.emit('temp', data={'value1':j})
        j+=1
        socketio.sleep(0)

@socketio.on('check2')
def gen2(data):
    j= 10
    while True:
        socketio.emit('temp2', data={'value2':j})
        j+=1
        socketio.sleep(3)

if __name__== "__main__":
    socketio.run(app,debug=True, host='127.0.0.1', port=5000) 
```

**index.html:**

```
<!DOCTYPE html>
<html lang="en">
<head>
    <title>websocket</title>
</head>
<body>
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/socket.io/1.7.3/socket.io.min.js"></script>
    <script type="text/javascript">
    $(document).ready(function(){
     var socket = io.connect('http://' + document.domain + ':' + location.port);

   socket.on('connect' ,function()
     {
      socket.emit('check' ,{data: 'User Connected'})
     });

    socket.on('temp' ,function(msg) 
    {
        console.log(msg.value1)

        $('#test').html('<p> Value 1: ' + msg.value1 + '</p>');
     });
     socket2.on('temp2' ,function(msg) 
     {
        console.log(msg.value2)
        $('#test2').html('<p> Value 2: ' + msg.value2 + '</p>');
     });
    });

 </script>
 <p id='test'> Hello </p>
 <br/><br/>
 <p id='test2'> Hello </p>

</body>
</html>
```
I am relatively new to flask websockets and any kind of help is appreciated.

**Comment by miguelgrinberg** [comment, marked as answer]
I'm sorry, I don't understand. You are emitting `temp` events as fast as your server can, and you are emitting `temp2` every 3 seconds. How can the updates happen at the same time when the rates are different?

This may be more a question about front end development, because I don't see how this can be solved from the Socket.IO side unless you combine your two data items into a single event.

**Reply by athulvis** [reply]
Single event makes the update slower so that it will take 3 seconds to update both values.

Is it possible to update value 1 with no delay and value 2 with 3 seconds delay in parallel?

Also I checked the logs. The flask-socketio server is sending only  `value1` to the server, as seen in the log below:

```
emitting event "temp" to all [/]
286b7a1d1ca74b5897d898688cfedf71: Sending packet MESSAGE data 2["temp",{"value1":35147}]
emitting event "temp" to all [/]
286b7a1d1ca74b5897d898688cfedf71: Sending packet MESSAGE data 2["temp",{"value1":35148}]
emitting event "temp" to all [/]
286b7a1d1ca74b5897d898688cfedf71: Sending packet MESSAGE data 2["temp",{"value1":35149}]
emitting event "temp" to all [/]
```
Its not emitting `value2` along with `value1` considering the delay time.

**Reply by miguelgrinberg** [reply]
Wouldn't you need something in the client that emits `check2` for the value2 updates to start flowing?

**Reply by athulvis** [reply]
Thanks for noticing it. There is a possibility of having issue in client side. Let me check that one.

**Reply by athulvis** [reply]
@miguelgrinberg Thanks for the insight. 

I tried to initiate a new socket on client side. But at a time, only one socket was initiated. So I used asyncio and now both sockets are working.

`Updated index.html` (removed multiple sockets after suggestion from @miguelgrinberg )

```
<!DOCTYPE html>
<html lang="en">
<head>
	<title> flask websocket</title>
    
</head>
<body>
	<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/socket.io/1.7.3/socket.io.min.js"></script>
    <script type="text/javascript">
    $(document).ready(function(){
     var socket = io.connect('http://' + document.domain + ':' + location.port);

   socket.on('connect' ,async function()
     {
     await socket.emit('check' ,{data: 'User Connected'})
     });

    socket.on('temp' , async function(msg) 
   {
      await $('#test').html('<p> Value 1: ' + msg.value1 + '</p>');
      console.log(msg.value1)
  });

       socket.on('temp2' ,async function(msg) 
     {
        await $('#test2').html('<p> Value 2: ' + msg.value2 + '</p>');
        console.log(msg.value2)
     });
    });

 </script>
 <br/><br/>
 <p id='test'> Hello </p>
 <br/><br/>
 <p id='test2'> Hello </p>

</body>
</html>
```

**Reply by miguelgrinberg** [reply]
You can emit twice from the same socket. Don't create two connections.

**Reply by athulvis** [reply]
Thanks. That worked too, Updated above answer.
